README.md
## Custom signature

This package project offers an easy way to get your signature with a custom UI, and rotates automatically into landscape mode for most comfortable space to sign.

## Usage

```dart
CustomSignature(
    context: context
    title: 'Signature',
    closeButtonColor: Colors.indigo,
    icon: FontAwesomeIcons.circleXmark,
    enabledButtonColor: Colors.green,
    disabledButtonColor: Colors.lightGreen.withOpacity(0.5),
    penColor: Colors.black,
    deleteButtonText: 'Delete',
    doneButtonText: 'Done',
    titleStyle: const TextStyle(fontWeight: FontWeight.w400, fontSize: 16),
    buttonTextStyle: const TextStyle(fontSize: 16, fontWeight: FontWeight.w600),
    // optional properties
    backgroundColor: Colors.white,
    textColor: Colors.black,
    closeAlert: AlertDialog(
        content:
            const Text('Do you want to exit?'),
        actions: [
        ElevatedButton(
            child: const Text('Ok'),
            onPressed: () {
                Navigator.pop(context);
                Navigator.pop(context);
            }),
        MaterialButton(
            onPressed: () => Navigator.pop(context),
            child: const Text('No')),
        ],
    ),
);
```

## Additional information

Is recommended that the screen to going to navigate has Stateful Widget with preferred orientation configured, to turn back the default orientation (portrait).

```dart
SystemChrome.setPreferredOrientations([
      DeviceOrientation.portraitUp,
      DeviceOrientation.portraitDown,
    ]);
  }
```

CHANGELOG.md
## 2.0.0
* Add title style and text button style

## 1.0.4
* Add screen util tool to adjust to screen size.
* Add safe margins and minor UI adjustments.

## 1.0.3
* Size adjustments.

## 1.0.2
* Remove unnesessary widget.

## 1.0.1
* Example updated.

## 1.0.0
* Remove nextPage widget parameter and navigate backward with a pop() function.


## 0.0.5
* UI optimizations.

## 0.0.4
* UI optimizations.

## 0.0.3
* Package adjustments.
* Example updated.
* Github repo added.

## 0.0.2
* Package adjustements.
* Example updated.

## 0.0.1 First release
* Initial version.
* Example and documentation added.


lib/custom_signature.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';
import 'package:signature/signature.dart';

/// Main class
/// [closeAlert], [backgroundColor] and [textColor] are optional properties. Provides an alert to show when user wants to exit: sets the background color in signature area; and sets the title color, respectively.
class CustomSignature extends StatefulWidget {
  final BuildContext context;
  final String title;
  final Color closeButtonColor;
  final Color enabledButtonColor;
  final Color disabledButtonColor;
  final Color penColor;
  final IconData icon;
  final String deleteButtonText;
  final String doneButtonText;
  final TextStyle titleStyle;
  final TextStyle buttonTextStyle;
  final Widget? closeAlert;
  final Color? backgroundColor;
  final Color? textColor;

  /// Constructor
  const CustomSignature({
    super.key,
    required this.context,
    required this.title,
    required this.closeButtonColor,
    required this.enabledButtonColor,
    required this.disabledButtonColor,
    required this.penColor,
    required this.icon,
    required this.deleteButtonText,
    required this.doneButtonText,
    required this.titleStyle,
    required this.buttonTextStyle,
    this.closeAlert,
    this.backgroundColor = Colors.transparent,
    this.textColor = Colors.black,
  });

  @override
  State<CustomSignature> createState() => _CustomSignatureState();
}

class _CustomSignatureState extends State<CustomSignature> {
  /// Sets controller.
  late SignatureController _controller;

  /// Set active or inactive the buttons to change their colors.
  bool isButtonActive = false;

  @override
  void initState() {
    super.initState();

    /// Initialize controller
    _controller = SignatureController(
      penStrokeWidth: 2,
      penColor: widget.penColor,
      exportBackgroundColor: Colors.grey[400],
      exportPenColor: Colors.black,
      onDrawStart: () {
        setState(() => isButtonActive = true);
      },
      onDrawEnd: () {},
    );

    _controller.addListener(() {});

    /// Sets screen in landscape mode when that appears.
    SystemChrome.setPreferredOrientations([
      DeviceOrientation.landscapeRight,
      DeviceOrientation.landscapeLeft,
    ]);
  }

  @override
  void dispose() {
    super.dispose();

    _controller.dispose();

    /// Sets screen in portrait mode when that disappears.
    SystemChrome.setPreferredOrientations([
      DeviceOrientation.portraitUp,
      DeviceOrientation.portraitDown,
    ]);
  }

  @override
  Widget build(BuildContext context) {
    ScreenUtil.init(context, designSize: const Size(812, 375));
    context = widget.context;

    return Scaffold(
      backgroundColor: Colors.white,
      bottomNavigationBar: _bottomActions(),
      body: SafeArea(
        child: Column(
          children: [
            _headerInformation(widget.title),
            _signContainer(),
          ],
        ),
      ),
    );
  }

  /// Builds the header of the screen, that contains the title and the close button.
  Container _headerInformation(String title) {
    return Container(
      width: double.infinity,
      padding: EdgeInsets.only(left: 20.w, right: 22.w, top: 15.h),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          Text(title, style: widget.titleStyle),
          IconButton(
            icon: Icon(widget.icon, size: 23.w),
            color: widget.closeButtonColor,
            onPressed: () {
              if (widget.closeAlert == null) {
                Navigator.pop(context);
              } else {
                showDialog(
                    context: context, builder: (_) => widget.closeAlert!);
              }
            },
          ),
        ],
      ),
    );
  }

  /// Builds signature container
  Widget _signContainer() {
    return Expanded(
      child: Signature(
        controller: _controller,
        width: double.infinity,
        backgroundColor: widget.backgroundColor!,
      ),
    );
  }

  /// Builds in the bottom of the screen, the buttons to clean or send the signature, besides theirs properties.
  Widget _bottomActions() {
    return BottomAppBar(
        height: 85.h,
        elevation: 0,
        color: Colors.white,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Divider(
              height: 1,
              color: widget.disabledButtonColor,
              thickness: 1,
              indent: 0,
              endIndent: 0,
            ),
            Padding(
              padding: EdgeInsets.only(right: 30.w, top: 10.h),
              child: Row(mainAxisAlignment: MainAxisAlignment.end, children: [
                TextButton(
                  style: TextButton.styleFrom(
                      foregroundColor: widget.enabledButtonColor),
                  onPressed: (!isButtonActive)
                      ? null
                      : () {
                          setState(() {
                            isButtonActive = false;
                            _controller.clear();
                          });
                        },
                  child: SizedBox(
                    width: 142.w,
                    child: Text(
                      widget.deleteButtonText,
                      textAlign: TextAlign.center,
                      style: widget.buttonTextStyle.copyWith(
                        color: (isButtonActive)
                            ? widget.enabledButtonColor
                            : widget.disabledButtonColor,
                      ),
                    ),
                  ),
                ),
                SizedBox(width: 10.w),
                ElevatedButton(
                  style: ElevatedButton.styleFrom(
                      elevation: 0,
                      disabledBackgroundColor: widget.disabledButtonColor,
                      backgroundColor: widget.enabledButtonColor),
                  onPressed: (!isButtonActive) ? null : () => exportImage(),
                  child: SizedBox(
                    width: 212.w,
                    child: Text(
                      widget.doneButtonText,
                      textAlign: TextAlign.center,
                      style: widget.buttonTextStyle.copyWith(
                        color: (isButtonActive)
                            ? Colors.white
                            : widget.enabledButtonColor,
                      ),
                    ),
                  ),
                ),
              ]),
            ),
          ],
        ));
  }

  /// Exports image in 8-bit format and send it to the next sceen.
  Future<void> exportImage() async {
    final Uint8List? data =
        await _controller.toPngBytes(height: 1000, width: 1000);
    if (data == null) {
      return;
    }

    if (!mounted) return;
    Navigator.pop(context, data);
  }
}


example/main.dart
import 'package:custom_signature/custom_signature.dart';
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
        debugShowCheckedModeBanner: false, title: 'Example', home: FirstPage());
  }
}

class FirstPage extends StatelessWidget {
  const FirstPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Center(
      child: ElevatedButton(
          child: const Text('To sign'),
          onPressed: () async {
            await Navigator.push(
                context,
                MaterialPageRoute(
                    builder: (BuildContext context) => CustomSignature(
                          context: context,
                          title: 'Signature',
                          closeButtonColor: Colors.indigo,
                          icon: Icons.close,
                          enabledButtonColor: Colors.green,
                          disabledButtonColor:
                              Colors.lightGreen.withOpacity(0.5),
                          penColor: Colors.black,
                          titleStyle: const TextStyle(
                              fontWeight: FontWeight.w400, fontSize: 16),
                          buttonTextStyle: const TextStyle(
                              fontSize: 16, fontWeight: FontWeight.w600),
                          textColor: Colors.black,
                          deleteButtonText: 'Delete',
                          doneButtonText: 'Done',
                          closeAlert: AlertDialog(
                            content: const Text('Do you want to exit?'),
                            actions: [
                              ElevatedButton(
                                  child: const Text('Ok'),
                                  onPressed: () {
                                    Navigator.pop(context);
                                    Navigator.pop(context);
                                  }),
                              MaterialButton(
                                  onPressed: () => Navigator.pop(context),
                                  child: const Text('No')),
                            ],
                          ),
                        ))).then((value) {
              showDialog(
                  context: context,
                  builder: (_) {
                    return AlertDialog(
                      actions: [
                        TextButton(
                            onPressed: () => Navigator.pop(context),
                            child: const Text('Close'))
                      ],
                      title: const Text('Signature'),
                      content: Image.memory(value),
                    );
                  });
            });
          }),
    ));
  }
}
